mount/mount_linux.go

```diff
@@ -363,24 +364,29 @@ func mountAt(chdir string, source, target, fstype string, flags uintptr, data st
 		return unix.Mount(source, target, fstype, flags, data)
 	}
 
-	f, err := os.Open(chdir)
-	if err != nil {
-		return fmt.Errorf("failed to mountat: %w", err)
-	}
-	defer f.Close()
+	ch := make(chan error, 1)
+	go func() {
+		runtime.LockOSThread()
 
-	fs, err := f.Stat()
-	if err != nil {
-		return fmt.Errorf("failed to mountat: %w", err)
-	}
+		// Do not unlock this thread.
```

Comment:
The locked thread could be main thread.
If so, when the goroutine exits, the main thread won't exit and it seems that go-runtime never schedules goroutine on this thread.(Shared information discussed offline)

https://github.com/golang/go/blob/947091d31ccda14b0a362adff37b6e037f0f59f3/src/runtime/proc.go#L1463-L1468

But the go-doc doesn't mention that. Maybe we can confirm that with go community.
If it is safe, this patch will be great!!!!!!!!!!!

Reply:
It is not a problem for the Go runtime. We have signaled to the Go runtime that we do not want any other goroutines to be scheduled onto the locked thread as the thread's kernel state has been dirtied, and the runtime arranges for that, irrespective of which thread the goroutine got locked to. [The runtime will spawn a new thread as needed to replace the wedged main thread.](https://pkg.go.dev/runtime/debug#SetMaxThreads) There were problems before go1.10, [but these have long been fixed.](https://www.weave.works/blog/linux-namespaces-golang-followup) The discussions on the golang/go issues linked from that article clearly demonstrate that the Go maintainers have improved the runtime behaviours around thread locking expressly to support use-cases exactly like this.

The main thread not exiting could be problematic in cases where the locked thread would need to exit immediately in order to release kernel resources such as unshared namespaces. As `unshare(CLONE_FS)` only unshares [a tiny amount of kernel state](https://elixir.bootlin.com/linux/v6.0/source/include/linux/fs_struct.h#L9) and there is only one main thread per process, wedging the main thread would not be a problem in this case.

There are other subtleties around altering the main thread's kernel state in the first place as `/proc/self/` and `/proc/[pid]/` reflects the kernel state of the respective process' main thread. In this case it is of trivial consequence; the current working directory for other goroutines (`os.Getwd()`) would be desynchronized from `/proc/self/cwd` (but not `/proc/thread-self/cwd` or `/proc/self/task/[tid]/cwd`).

If, in other cases, modifying the kernel state of the main thread would be problematic, there is a surefire way to prevent the main thread from getting locked to arbitrary goroutines: lock it to `func main()` instead!
```go
func init() {
	runtime.LockOSThread()
}
```
The process is terminated when `func main()` returns so the runtime would never have to wedge the main thread.

Reply:
FWIW I also tested this with GOMAXPROCS=1 and it did not have any issue.

Reply:
@corhere Yeah. Lock main thread will be good!
Thanks for the detail.